### PR TITLE
Java: Use Java 8 list.removeIf in place of Iterator #8214

### LINKED ISSUE #8214: Java: Use Java 8 list.removeIf in place of Iterator
Context: Someone wants to "filter" a list according to a given condition.

Pre-Java 8:
```java
Iterator<String> iter = list.iterator();
while (iter.hasNext()) {
    String entry = iter.next();
    if (condition) {
        iter.remove();
    }
}
```
Java 8:
```java
list.removeIf(entry -> {
    return condition;
});
```
There are more than 10 places where this can be applied.

### PR DESCRIPTION
**Pull request by Haozhe321**
Fixes #8214 

**Outline of Solution**

Issue is resolved by replacing the old method of using iterator to remove an element in a Collection i.e. ```iter.remove()```, to the new ```Collection.removeIf()``` method.

There are a total of 4 files that ```iter.remove()``` appeared in, and all of them have been replaced. 

For a trivial fix from 
```
Iterator<FeedbackResponseAttributes> iterResponse = responses.iterator();
while (iterResponse.hasNext()) {
    FeedbackResponseAttributes response = iterResponse.next();
    if (response.recipient.equals(instructorEmail)) {
        iterResponse.remove();
    }
}
```
to
```
responses.removeIf(response -> response.recipient.equals(instructorEmail));
```
Intellij provides an automatic refactoring option. 

For more complex cases where there are more conditions and variables defined in the `while` loop, I have used manual refactoring. 

**Note**
Note that [variables used in Lambda should be final or effectively final](http://www.javahabit.com/2016/06/16/understanding-java-8-lambda-final-finally-variable), hence in one of the fix  an array of 1 slot is created, as per suggested by Intellij (and the abovementioned website) to overcome this restriction.  For example, 
```int filteredResultsSize = totalResultsSize;``` 
has been changed to 
```final int[] filteredResultsSize = {totalResultsSize};```

### CODE REVIEW

File: src/main/java/teammates/storage/search/FeedbackResponseCommentSearchDocument.java [hunk 1]
```diff
@@ -465,22 +465,20 @@ private static int filterFeedbackResponseCommentResults(
         Iterator<Entry<String, List<FeedbackResponseAttributes>>> iterFr =
                 frCommentSearchResults.responses.entrySet().iterator();
 
-        int filteredResultsSize = totalResultsSize;
+        final int[] filteredResultsSize = {totalResultsSize};
         while (iterFr.hasNext()) {
```
[the remark this answers is not in the document]
Comment (Haozhe321) [reply src/main/java/teammates/storage/search/FeedbackResponseCommentSearchDocument.java]: ```frCommentSearchResults.responses.entrySet().forEach``` will work as well, in which case the resulting code looks like (ignore the unchanged ```isVisibleResponse``` for now)
```
frCommentSearchResults.responses.entrySet().forEach(fr -> {
    List<FeedbackResponseAttributes> frs = fr.getValue();

    frs.removeIf(response -> {
        InstructorAttributes instructor = getInstructorForCourseId(response.courseId, instructors);

        boolean isVisibleResponse = true;
        boolean isNotAllowedForInstructor =
                instructor == null
                        || !instructor.isAllowedForPrivilege(
                        response.giverSection, response.feedbackSessionName,
                        Const.ParamsNames.INSTRUCTOR_PERMISSION_VIEW_SESSION_IN_SECTIONS)
                        || !instructor.isAllowedForPrivilege(
                        response.recipientSection, response.feedbackSessionName,
                        Const.ParamsNames.INSTRUCTOR_PERMISSION_VIEW_SESSION_IN_SECTIONS);

        if (isNotAllowedForInstructor) {
            isVisibleResponse = false;
        }
        if (!isVisibleResponse) {
            int sizeOfCommentList = frCommentSearchResults.comments.get(response.getId()).size();
            filteredResultsSize[0] -= sizeOfCommentList;
            // TODO: also need to decrease the size for (fr)CommentSearchResults|studentSearchResults
            frCommentSearchResults.comments.remove(response.getId());
        }
        return !isVisibleResponse;
    });
});
```

The other choice would be as you suggested, 
```
frCommentSearchResults.responses.forEach((key, frs) -> frs.removeIf(response -> {
    InstructorAttributes instructor = getInstructorForCourseId(response.courseId, instructors);

    boolean isVisibleResponse = true;
    boolean isNotAllowedForInstructor =
            instructor == null
                    || !instructor.isAllowedForPrivilege(
                    response.giverSection, response.feedbackSessionName,
                    Const.ParamsNames.INSTRUCTOR_PERMISSION_VIEW_SESSION_IN_SECTIONS)
                    || !instructor.isAllowedForPrivilege(
                    response.recipientSection, response.feedbackSessionName,
                    Const.ParamsNames.INSTRUCTOR_PERMISSION_VIEW_SESSION_IN_SECTIONS);

    if (isNotAllowedForInstructor) {
        isVisibleResponse = false;
    }
    if (!isVisibleResponse) {
        int sizeOfCommentList = frCommentSearchResults.comments.get(response.getId()).size();
        filteredResultsSize[0] -= sizeOfCommentList;
        // TODO: also need to decrease the size for (fr)CommentSearchResults|studentSearchResults
        frCommentSearchResults.comments.remove(response.getId());
    }
    return !isVisibleResponse;
}));
```

Which one is better?

[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/main/java/teammates/storage/search/FeedbackResponseCommentSearchDocument.java]: I like the second one better; more compact.

**Comment by Haozhe321** [comment]
@wkurniawan07 Changes made!

**Comment by Haozhe321** [comment]
@wkurniawan07 Changes made!

**Pull request merged**